# Atividade prática com datasets de energia

Esse notebook contém a resolução da Etapa B (Python / Pandas) dos Datasets 1 ao 6


# **Dataset 1 — Appliances Energy Prediction**

**Situação:** identificar períodos de consumo elevado de eletrodomésticos e observar as condições de temperatura e umidade nesses momentos.

### 1. Carregar a amostra e inspecionar

In [ ]:
import pandas as pd

# Troque pelo nome do arquivo exportado do Orange (Save Data)
df1 = pd.read_csv('/content/Amostra_base_Appliances_Energy_Prediction.csv')

df1.head()

,Appliances,lights,T1,T2,T3,RH_1,RH_2,RH_3
0,40,0,20.8900,17.7600,20.2900,35.4000,39.1633,36.9000
1,90,10,21.8900,21.2900,21.6333,53.1000,45.3600,49.2267
2,50,0,21.3900,17.6333,21.6667,35.5000,40.5300,35.2000
3,50,0,21.3900,23.8900,22.0333,41.0333,34.8400,36.9333
4,70,0,19.9633,16.4633,20.0000,35.1267,40.1267,36.4000


### 2. Renomear colunas

Renomeamos a coluna `Appliances` para `Consumo_Eletrodomesticos` e simplificamos pelo menos três atributos ambientais.

In [ ]:
df1.rename(columns={
    'Appliances': 'Consumo_Eletrodomesticos',
    'lights': 'Luzes',
    'T1': 'Temp_Cozinha',
    'RH_1': 'Umid_Cozinha',
    'T2': 'Temp_Sala',
    'RH_2': 'Umid_Sala',
    'T3': 'Temp_Lavanderia',
    'RH_3': 'Umid_Lavanderia',
}, inplace=True)

df1.columns

Index(['Consumo_Eletrodomesticos', 'Luzes', 'Temp_Cozinha', 'Temp_Sala',
       'Temp_Lavanderia', 'Umid_Cozinha', 'Umid_Sala', 'Umid_Lavanderia'],
      dtype='object')

### 3. Maior consumo de eletrodomésticos registrado

In [ ]:
consumo_max1 = df1['Consumo_Eletrodomesticos'].max()
consumo_max1

770

### 4. Limiar de 70% do máximo e DataFrame de consumo elevado

In [ ]:
limiar1 = 0.70 * consumo_max1
df1_alto_consumo = df1[df1['Consumo_Eletrodomesticos'] > limiar1]
limiar1

539.0

### 5. Contagem e percentual da amostra

In [ ]:
qtd1 = df1_alto_consumo.shape[0]
pct1 = qtd1 / df1.shape[0] * 100
print(f'Registros acima do limiar: {qtd1}')
print(f'Percentual da amostra: {pct1:.2f}%')

Registros acima do limiar: 22
Percentual da amostra: 1.11%


### 6. Temperatura média (T1) e segundo DataFrame (consumo alto E temperatura acima da média)

In [ ]:
temp_media1 = df1['Temp_Cozinha'].mean()

df1_alto_consumo_e_temp = df1[
    (df1['Consumo_Eletrodomesticos'] > limiar1) &
    (df1['Temp_Cozinha'] > temp_media1)
]

print(f'Temperatura média (Temp_Cozinha): {temp_media1:.2f}')
print(f'Registros com consumo alto E temperatura acima da média: {df1_alto_consumo_e_temp.shape[0]}')

Temperatura média (Temp_Cozinha): 21.66
Registros com consumo alto E temperatura acima da média: 8


### 7. Comparação entre os dois DataFrames

A maior parte dos picos de consumo de energia dos eletrodomésticos (64% dos casos analisados) ocorreu sem que a cozinha estivesse com temperaturas acima da média. Isso indica que o alto consumo nesta residência não é provocado pelo calor do ambiente, mas provavelmente por outros fatores, como hábitos dos moradores, horários específicos ou o uso de vários aparelhos ao mesmo tempo.

Ademais, ao aplicar dois filtros simultâneos (alto consumo + alta temperatura), o número de registros diminuiu drasticamente (de 22 para apenas 8), confirmando que a adição de regras reduz ou mantém o tamanho da amostra, jamais o aumenta.

In [ ]:
print(f'Somente consumo alto: {df1_alto_consumo.shape[0]} registros')
print(f'Consumo alto + temperatura acima da média: {df1_alto_consumo_e_temp.shape[0]} registros')

Somente consumo alto: 22 registros
Consumo alto + temperatura acima da média: 8 registros



# **Dataset 2 — Steel Industry Energy Consumption**

**Situação:** localizar cenários de consumo elevado na indústria siderúrgica e verificar se coincidem com carga elevada ou fator de potência desfavorável.

### 1. Carregar a amostra e renomear colunas

In [ ]:
df2 = pd.read_csv('/content/Amostra_base_Steel_Industry_Energy_Consumption.csv')

df2 = df2.rename(columns={
    'Usage_kWh': 'Consumo_kWh',
    'Lagging_Current_Reactive.Power_kVarh': 'Pot_Reativa_Atraso',
    'Leading_Current_Reactive_Power_kVarh': 'Pot_Reativa_Adiantada',
    'Lagging_Current_Power_Factor': 'Fator_Potencia_Atraso',
    'Leading_Current_Power_Factor': 'Fator_Potencia_Adiantado',
})

df2.columns

Index(['date', 'Consumo_kWh', 'Pot_Reativa_Atraso', 'Pot_Reativa_Adiantada',
       'Fator_Potencia_Atraso', 'Fator_Potencia_Adiantado', 'Load_Type',
       'WeekStatus', 'Day_of_week'],
      dtype='object')

### 2. Inspeção inicial

In [ ]:
df2.head()

,date,Consumo_kWh,Pot_Reativa_Atraso,Pot_Reativa_Adiantada,Fator_Potencia_Atraso,Fator_Potencia_Adiantado,Load_Type,WeekStatus,Day_of_week
0,29/04/2018 07:15,2.88,3.82,0.0,60.20,100.00,Light_Load,Weekend,Sunday
1,04/10/2018 12:00,60.77,48.02,0.0,78.46,100.00,Maximum_Load,Weekday,Thursday
2,26/01/2018 11:30,120.42,59.65,0.0,89.61,100.00,Maximum_Load,Weekday,Friday
3,02/06/2018 14:30,3.13,0.00,16.6,100.00,18.53,Light_Load,Weekend,Saturday
4,07/12/2018 15:00,58.86,20.99,0.0,94.19,100.00,Medium_Load,Weekday,Friday


In [ ]:
#df2.shape
print(f'Total de registros: {df2.shape[0]}')
print(f'Total de colunas: {df2.shape[1]}')

Total de registros: 7008
Total de colunas: 9


In [ ]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7008 entries, 0 to 7007
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   date                      7008 non-null   object 
 1   Consumo_kWh               7008 non-null   float64
 2   Pot_Reativa_Atraso        7008 non-null   float64
 3   Pot_Reativa_Adiantada     7008 non-null   float64
 4   Fator_Potencia_Atraso     7008 non-null   float64
 5   Fator_Potencia_Adiantado  7008 non-null   float64
 6   Load_Type                 7008 non-null   object 
 7   WeekStatus                7008 non-null   object 
 8   Day_of_week               7008 non-null   object 
dtypes: float64(5), object(4)
memory usage: 492.9+ KB


In [ ]:
df2.describe()

,Consumo_kWh,Pot_Reativa_Atraso,Pot_Reativa_Adiantada,Fator_Potencia_Atraso,Fator_Potencia_Adiantado
count,7008.000000,7008.000000,7008.000000,7008.000000,7008.000000
mean,27.790164,13.206839,3.803711,80.672277,84.491448
std,33.717719,16.444246,7.356412,18.869855,30.314798
min,2.450000,0.000000,0.000000,37.870000,12.710000
25%,3.200000,2.380000,0.000000,63.602500,99.710000
50%,4.570000,5.040000,0.000000,87.990000,100.000000
75%,52.107500,22.977500,1.857500,98.842500,100.000000
max,153.140000,87.700000,27.650000,100.000000,100.000000


### 3. Maior consumo registrado e limiar de 75%

In [ ]:
consumo_max2 = df2['Consumo_kWh'].max()
limiar2 = 0.75 * consumo_max2
print(f'Consumo máximo: {consumo_max2}')
print(f'Limiar (75%): {limiar2:.2f}')

Consumo máximo: 153.14
Limiar (75%): 114.85


### 4. DataFrame de consumo elevado, quantidade e percentual

In [ ]:
df2_alto_consumo = df2[df2['Consumo_kWh'] > limiar2]

qtd2 = df2_alto_consumo.shape[0]
pct2 = qtd2 / df2.shape[0] * 100
print(f'Registros acima do limiar: {qtd2}')
print(f'Percentual da amostra: {pct2:.2f}%')

Registros acima do limiar: 129
Percentual da amostra: 1.84%


### 5. Registros que pertencem a 'Maximum Load'

In [ ]:
qtd_maximum_load = (df2_alto_consumo['Load_Type'] == 'Maximum_Load').sum()
print(f'Registros de consumo elevado com Load_Type = Maximum_Load: {qtd_maximum_load}')
print(f'Percentual dentro do grupo de consumo elevado: {qtd_maximum_load / qtd2 * 100:.2f}%')

Registros de consumo elevado com Load_Type = Maximum_Load: 62
Percentual dentro do grupo de consumo elevado: 48.06%


### 6. Explorar o fator de potência e definir um limite para valores baixos

In [ ]:
df2['Fator_Potencia_Atraso'].describe()

,Fator_Potencia_Atraso
count,7008.000000
mean,80.672277
std,18.869855
min,37.870000
25%,63.602500
50%,87.990000
75%,98.842500
max,100.000000


Observe o `describe()` acima (mín, quartis, média) e escolha um limite coerente para "fator de potência baixo" — por exemplo, um valor próximo do 1º quartil ou um valor de referência técnico (fatores de potência abaixo de 0.85–0.90 costumam ser considerados baixos). Ajuste o valor abaixo conforme os dados observados.

In [ ]:
limite_fator_potencia = 65  # baseado no 1º quartil (63.60) do describe() acima
limite_fator_potencia

65

### 7. DataFrame com consumo elevado E fator de potência abaixo do limite

In [ ]:
df2_alto_consumo_fp_baixo = df2[
    (df2['Consumo_kWh'] > limiar2) &
    (df2['Fator_Potencia_Atraso'] < limite_fator_potencia)
]

qtd3 = df2_alto_consumo_fp_baixo.shape[0]
pct3 = qtd3 / df2.shape[0] * 100
print(f'Registros com consumo elevado e fator de potência baixo: {qtd3}')
print(f'Percentual da amostra: {pct3:.2f}%')

Registros com consumo elevado e fator de potência baixo: 0
Percentual da amostra: 0.00%


### 8. Por que esse segundo conjunto merece mais atenção?

Dos 129 registros (1,84%) com consumo acima de 75% do máximo, 62 (48,06%) já eram de carga máxima (Maximum_Load), o que é coerente, já que picos de consumo tendem a coincidir com esse tipo de carga. Ainda assim, nenhum desses registros apresentou fator de potência abaixo de 65 ao mesmo tempo, indicando que, mesmo nos momentos de maior demanda, a eficiência elétrica se manteve em níveis aceitáveis.

Portanto, esse cruzamento continua relevante como indicador de acompanhamento: se em análises futuras ele deixar de ser zero, sinalizará ineficiência elétrica justamente nos momentos mais custosos para a operação, ou seja, quando o impacto financeiro e o desgaste dos equipamentos são maiores.



# **Dataset 3 — Power Consumption of Tetouan City**

**Situação:** identificar qual das três zonas de distribuições de energia em Tetouan apresenta o maior pico de consumo e observar as condições ambientais registradas nos momentos de maior demanda

### 1. Carregar a amostra e inspecionar

In [ ]:
url3 = '/content/Amostra_base_Tetouan_City.csv'

df3 = pd.read_csv(url3)

df3.columns = (
    df3.columns
       .str.replace(r'\s+', ' ', regex=True)
       .str.strip()
)

df3.head()

,DateTime,Temperature,Humidity,Wind Speed,Zone 1 Power Consumption,Zone 2 Power Consumption,Zone 3 Power Consumption
0,4/18/2017 16:20,22.76,29.73,0.085,32985.14532,19396.7,17623.3
1,12/27/2017 17:10,15.92,54.44,0.085,34737.64259,29803.0,14417.3
2,1/2/2017 11:20,15.08,57.23,0.076,27894.68354,20808.5,16458.8
3,3/19/2017 20:20,12.42,72.30,0.081,42476.93617,23634.1,25089.7
4,4/24/2017 5:50,14.58,80.60,0.080,21341.14101,12006.1,11531.6


In [ ]:
print(f'Total de registros: {df3.shape[0]}')
print(f'Total de colunas: {df3.shape[1]}')

Total de registros: 7863
Total de colunas: 7


In [ ]:
df3.info()
df3.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7863 entries, 0 to 7862
Data columns (total 7 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   DateTime                  7863 non-null   object 
 1   Temperature               7863 non-null   float64
 2   Humidity                  7863 non-null   float64
 3   Wind Speed                7863 non-null   float64
 4   Zone 1 Power Consumption  7863 non-null   float64
 5   Zone 2 Power Consumption  7863 non-null   float64
 6   Zone 3 Power Consumption  7863 non-null   float64
dtypes: float64(6), object(1)
memory usage: 430.1+ KB


,Temperature,Humidity,Wind Speed,Zone 1 Power Consumption,Zone 2 Power Consumption,Zone 3 Power Consumption
count,7863.000000,7863.000000,7863.000000,7863.000000,7863.000000,7863.000000
mean,18.789376,68.400612,1.965602,32371.549651,21060.733609,17845.600069
std,5.864564,15.665773,2.350954,7134.686966,5202.357437,6573.203996
min,3.541000,13.340000,0.054000,14612.658230,9436.250000,6211.760000
25%,14.340000,58.485000,0.078000,26263.115455,17013.400000,13200.700000
50%,18.690000,70.000000,0.086000,32249.429660,20851.600000,16490.300000
75%,22.820000,81.500000,4.915000,37421.355555,24812.200000,21763.600000
max,39.600000,94.700000,6.325000,51571.498340,36437.200000,47441.700000


### 2. Renomear as colunas de consumo

In [ ]:
df3 = df3.rename(columns={
    'Zone 1 Power Consumption': 'Consumo_Zona_1',
    'Zone 2 Power Consumption': 'Consumo_Zona_2',
    'Zone 3 Power Consumption': 'Consumo_Zona_3'
})

df3.columns

Index(['DateTime', 'Temperature', 'Humidity', 'Wind Speed', 'Consumo_Zona_1',
       'Consumo_Zona_2', 'Consumo_Zona_3'],
      dtype='object')

### 3. Consumo máximo registrado em cada zona

In [ ]:
maximos3 = df3[
    ['Consumo_Zona_1', 'Consumo_Zona_2', 'Consumo_Zona_3']
].max()

print('Consumo máximo de cada zona:')
print(maximos3)

Consumo máximo de cada zona:
Consumo_Zona_1    51571.49834
Consumo_Zona_2    36437.20000
Consumo_Zona_3    47441.70000
dtype: float64


### 4. Identificar a zona com o maior pico de consumo

In [ ]:
zona_maior3 = maximos3.idxmax()
maior_pico3 = maximos3.max()

print(f'Zona com o maior pico: {zona_maior3}')
print(f'Maior consumo registrado: {maior_pico3:.2f}')

Zona com o maior pico: Consumo_Zona_1
Maior consumo registrado: 51571.50


### 5. Limiar de 70% e DataFrame de consumo elevado

In [ ]:
limiar3 = 0.70 * maior_pico3

df3_alto_consumo = df3[
    df3[zona_maior3] > limiar3
]

print(f'Limiar de 70%: {limiar3:.2f}')
df3_alto_consumo.head()

Limiar de 70%: 36100.05


,DateTime,Temperature,Humidity,Wind Speed,Consumo_Zona_1,Consumo_Zona_2,Consumo_Zona_3
3,3/19/2017 20:20,12.42,72.30,0.081,42476.93617,23634.1,25089.7
8,12/10/2017 17:50,14.66,65.92,0.075,37286.69202,30716.2,15834.8
12,6/5/2017 23:30,21.03,86.50,0.068,40993.90728,25383.4,24960.0
14,8/23/2017 17:50,27.06,47.58,4.905,37475.16093,24276.2,25140.6
19,6/22/2017 2:40,19.20,65.81,4.924,40676.02649,24754.7,25095.9


### 6. Quantidade e percentual de registros com consumo elevado

In [ ]:
qtd_alto3 = df3_alto_consumo.shape[0]
pct_alto3 = qtd_alto3 / df3.shape[0] * 100

print(f'Registros acima do limiar: {qtd_alto3}')
print(f'Percentual da amostra: {pct_alto3:.2f}%')

Registros acima do limiar: 2388
Percentual da amostra: 30.37%


### 7. Consumo elevado e temperatura acima da média

In [ ]:
temperatura_media3 = df3['Temperature'].mean()

df3_alto_consumo_temp = df3[
    (df3[zona_maior3] > limiar3) &
    (df3['Temperature'] > temperatura_media3)
]

print(f'Temperatura média: {temperatura_media3:.2f}')
print(
    'Registros com consumo elevado e temperatura acima da média:',
    df3_alto_consumo_temp.shape[0]
)

Temperatura média: 18.79
Registros com consumo elevado e temperatura acima da média: 1597


### 8. Comparação entre os dois DataFrames

In [ ]:
reducao3 = qtd_alto3 - df3_alto_consumo_temp.shape[0]
percentual_reducao3 = reducao3 / qtd_alto3 * 100

print(f'Somente consumo elevado: {qtd_alto3} registros')
print(
    'Consumo elevado e temperatura acima da média:',
    df3_alto_consumo_temp.shape[0],
    'registros'
)
print(f'Redução de registros: {reducao3}')
print(f'Redução percentual: {percentual_reducao3:.2f}%')

Somente consumo elevado: 2388 registros
Consumo elevado e temperatura acima da média: 1597 registros
Redução de registros: 791
Redução percentual: 33.12%


### 9. Interpretação

A Zona 1 apresentou o maior pico de consumo da amostra, com aproximadamente
51.571,50. Ao considerar como consumo elevado os valores acima de 70% desse
máximo, foram encontrados 2.388 registros, equivalentes a 30,37% da amostra.

Ao incluir também a condição de temperatura acima da média de 18,79, o número
de registros diminuiu para 1.597. Isso representa uma redução de 791 registros,
ou 33,12% do conjunto de consumo elevado.

A redução acontece porque o segundo DataFrame exige que duas condições sejam
verdadeiras simultaneamente: consumo alto na Zona 1 e temperatura acima da
média. O resultado mostra uma relação dentro da amostra analisada, mas não
permite afirmar que a temperatura seja a causa do aumento do consumo.

# **Dataset 4 — Solar Power Generation Data**

**Situação:** identificar períodos de alta geração de energia na usina fotovoltaica e verificar quais inversores aparecem com maior frequência nesses momentos.

### 1. Carregar a amostra e inspecionar

In [ ]:
url4 = '/content/Amostra_base_Solar_Power_Generation.csv'

df4 = pd.read_csv(url4)

df4.head()

,DATE_TIME,SOURCE_KEY,DC_POWER,AC_POWER,DAILY_YIELD,TOTAL_YIELD
0,10-06-2020 19:00,uHbuxQJl8lW7ozc,0.00,0.000,6645.000,7240508.000
1,28-05-2020 00:15,rGa61gmuvPhdLxV,0.00,0.000,0.000,7206549.000
2,17-05-2020 15:15,sjndEbLyjtCKgGv,7333.88,718.062,6442.000,7036055.000
3,01-06-2020 04:15,ih0vzX44oOqAx2f,0.00,0.000,0.000,6308232.000
4,17-05-2020 09:00,z9Y9gH1T5YWrNuG,7665.50,750.438,961.375,7021701.375


In [ ]:
print(f'Total de registros: {df4.shape[0]}')
print(f'Total de colunas: {df4.shape[1]}')

Total de registros: 13756
Total de colunas: 6


In [ ]:
df4.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13756 entries, 0 to 13755
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   DATE_TIME    13756 non-null  object 
 1   SOURCE_KEY   13756 non-null  object 
 2   DC_POWER     13756 non-null  float64
 3   AC_POWER     13756 non-null  float64
 4   DAILY_YIELD  13756 non-null  float64
 5   TOTAL_YIELD  13756 non-null  float64
dtypes: float64(4), object(2)
memory usage: 644.9+ KB


In [ ]:
df4.describe()

,DC_POWER,AC_POWER,DAILY_YIELD,TOTAL_YIELD
count,13756.000000,13756.000000,13756.000000,1.375600e+04
mean,3135.848662,306.695495,3319.235216,6.978829e+06
std,4023.069619,393.124284,3142.249927,4.162164e+05
min,0.000000,0.000000,0.000000,6.183645e+06
25%,0.000000,0.000000,0.000000,6.509779e+06
50%,370.071500,35.771400,2730.425000,7.146218e+06
75%,6350.562500,622.065500,6293.782500,7.267996e+06
max,14416.100000,1405.590000,9163.000000,7.846821e+06


### 2. Renomear as principais variáveis

In [ ]:
df4 = df4.rename(columns={
    'DC_POWER': 'Potencia_CC',
    'AC_POWER': 'Potencia_CA',
    'DAILY_YIELD': 'Geracao_Diaria',
    'TOTAL_YIELD': 'Geracao_Total'
})

df4.columns

Index(['DATE_TIME', 'SOURCE_KEY', 'Potencia_CC', 'Potencia_CA',
       'Geracao_Diaria', 'Geracao_Total'],
      dtype='object')

### 3. Maior potência CA registrada e limiar de 70%

In [ ]:
potencia_ca_max4 = df4['Potencia_CA'].max()
limiar4 = 0.70 * potencia_ca_max4

print(f'Maior potência CA: {potencia_ca_max4:.2f}')
print(f'Limiar de 70%: {limiar4:.2f}')

Maior potência CA: 1405.59
Limiar de 70%: 983.91


### 4. DataFrame de alta geração, quantidade e percentual

In [ ]:
df4_alta_geracao = df4[
    df4['Potencia_CA'] > limiar4
]

qtd_alta4 = df4_alta_geracao.shape[0]
pct_alta4 = qtd_alta4 / df4.shape[0] * 100

print(f'Registros de alta geração: {qtd_alta4}')
print(f'Percentual da amostra: {pct_alta4:.2f}%')

Registros de alta geração: 1185
Percentual da amostra: 8.61%


### 5. Frequência dos inversores nos registros de alta geração

In [ ]:
frequencia_inversores4 = (
    df4_alta_geracao['SOURCE_KEY']
    .value_counts()
)

frequencia_inversores4

,count
SOURCE_KEY,
adLQvlD726eNBSB,68
McdE0feGgRqW7Ca,65
ZnxXDlPa8U1GXgE,63
pkci93gMrogZuBj,62
ZoEaEvLYb1n2sOq,61
1IF53ai7Xc0U56Y,59
YxYtjZvoooNbGkE,58
sjndEbLyjtCKgGv,57
z9Y9gH1T5YWrNuG,56


### 6. Inversor mais frequente na alta geração

In [ ]:
inversor_mais_frequente4 = frequencia_inversores4.idxmax()
qtd_inversor4 = frequencia_inversores4.max()

pct_inversor4 = qtd_inversor4 / qtd_alta4 * 100

print(f'Inversor mais frequente: {inversor_mais_frequente4}')
print(f'Quantidade de aparições: {qtd_inversor4}')
print(
    f'Percentual dentro dos registros de alta geração: '
    f'{pct_inversor4:.2f}%'
)

Inversor mais frequente: adLQvlD726eNBSB
Quantidade de aparições: 68
Percentual dentro dos registros de alta geração: 5.74%


### 7. Interpretação

A maior potência CA registrada na amostra foi de 1.405,59. O limiar de alta
geração, calculado como 70% desse valor máximo, foi de 983,91.

Foram encontrados 1.185 registros acima desse limite, equivalentes a 8,61% da
amostra. Ao analisar a frequência dos inversores nesses momentos, o inversor
`adLQvlD726eNBSB` foi o mais frequente, com 68 aparições, representando 5,74%
dos registros de alta geração.

Esse resultado indica que esse inversor apareceu mais vezes no recorte
selecionado. Entretanto, a frequência isolada não permite concluir que ele seja
o inversor de melhor desempenho ou que os demais apresentem falhas. Para essa
conclusão, seria necessário considerar outros fatores, como o total de registros
de cada inversor, os horários de operação e as condições ambientais.

# **Dataset 5 — Wind & Solar Energy Production Dataset**

**Situação:** comparar a frequência de períodos de alta geração solar e de alta
geração eólica, considerando o máximo próprio de cada fonte.

### 1. Carregar a amostra e inspecionar

In [ ]:
url5 = '/content/Amostra_base_Wind_Solar_Energy_Production.csv'

df5 = pd.read_csv(url5)

df5.head()

,Date,Source,Start_Hour,End_Hour,Production
0,12/18/2022,Wind,21,22,11400
1,1/29/2024,Wind,14,15,7917
2,8/1/2024,Solar,15,16,8835
3,11/10/2020,Wind,1,2,989
4,12/19/2022,Wind,1,2,12526


In [ ]:
print(f'Total de registros: {df5.shape[0]}')
print(f'Total de colunas: {df5.shape[1]}')

Total de registros: 10373
Total de colunas: 5


In [ ]:
df5.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10373 entries, 0 to 10372
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Date        10373 non-null  object
 1   Source      10373 non-null  object
 2   Start_Hour  10373 non-null  int64 
 3   End_Hour    10373 non-null  int64 
 4   Production  10373 non-null  int64 
dtypes: int64(3), object(2)
memory usage: 405.3+ KB


In [ ]:
df5.describe()

,Start_Hour,End_Hour,Production
count,10373.000000,10373.000000,10373.000000
mean,11.530898,11.489733,6193.352164
std,6.947598,6.949144,3981.573520
min,0.000000,0.000000,91.000000
25%,6.000000,5.000000,3117.000000
50%,12.000000,11.000000,5334.000000
75%,18.000000,18.000000,8443.000000
max,23.000000,23.000000,22929.000000


### 2. Verificar as fontes de geração disponíveis

In [ ]:
df5['Source'].value_counts()

,count
Source,
Wind,8527
Solar,1846


### 3. Separar os dados solar e eólico e renomear as variáveis

In [ ]:
df5_solar = df5[df5['Source'] == 'Solar'].copy()
df5_eolica = df5[df5['Source'] == 'Wind'].copy()

df5_solar = df5_solar.rename(
    columns={'Production': 'Geracao_Solar'}
)

df5_eolica = df5_eolica.rename(
    columns={'Production': 'Geracao_Eolica'}
)

print(f'Registros solares: {df5_solar.shape[0]}')
print(f'Registros eólicos: {df5_eolica.shape[0]}')

Registros solares: 1846
Registros eólicos: 8527


### 4. Valor máximo de geração para cada fonte

In [ ]:
max_solar5 = df5_solar['Geracao_Solar'].max()
max_eolica5 = df5_eolica['Geracao_Eolica'].max()

print(f'Máximo da geração solar: {max_solar5}')
print(f'Máximo da geração eólica: {max_eolica5}')

Máximo da geração solar: 16316
Máximo da geração eólica: 22929


### 5. Limiares de 70% do máximo de cada fonte

In [ ]:
limiar_solar5 = 0.70 * max_solar5
limiar_eolica5 = 0.70 * max_eolica5

print(f'Limiar solar (70%): {limiar_solar5:.2f}')
print(f'Limiar eólico (70%): {limiar_eolica5:.2f}')

Limiar solar (70%): 11421.20
Limiar eólico (70%): 16050.30


### 6. DataFrames de alta geração solar e eólica

In [ ]:
df5_solar_alta = df5_solar[
    df5_solar['Geracao_Solar'] > limiar_solar5
]

df5_eolica_alta = df5_eolica[
    df5_eolica['Geracao_Eolica'] > limiar_eolica5
]

print('Exemplo de registros de alta geração solar:')
display(df5_solar_alta.head())

print('Exemplo de registros de alta geração eólica:')
display(df5_eolica_alta.head())

Exemplo de registros de alta geração solar:


,Date,Source,Start_Hour,End_Hour,Geracao_Solar
77,7/8/2025,Solar,13,14,13125
178,3/19/2025,Solar,13,14,13786
183,8/17/2025,Solar,12,13,11455
187,7/16/2025,Solar,12,13,16160
207,3/31/2025,Solar,12,13,12441


Exemplo de registros de alta geração eólica:


,Date,Source,Start_Hour,End_Hour,Geracao_Eolica
20,1/3/2024,Wind,0,1,17272
32,3/11/2021,Wind,13,14,16144
103,1/29/2025,Wind,0,1,18199
151,12/8/2024,Wind,6,7,16946
192,11/19/2024,Wind,2,3,16568


### 7. Quantidade e percentual de registros de alta geração

In [ ]:
qtd_solar_alta5 = df5_solar_alta.shape[0]
qtd_eolica_alta5 = df5_eolica_alta.shape[0]

pct_solar_alta5 = (
    qtd_solar_alta5 / df5_solar.shape[0] * 100
)

pct_eolica_alta5 = (
    qtd_eolica_alta5 / df5_eolica.shape[0] * 100
)

print(f'Registros de alta geração solar: {qtd_solar_alta5}')
print(f'Percentual dentro da geração solar: {pct_solar_alta5:.2f}%')

print(f'Registros de alta geração eólica: {qtd_eolica_alta5}')
print(f'Percentual dentro da geração eólica: {pct_eolica_alta5:.2f}%')

Registros de alta geração solar: 45
Percentual dentro da geração solar: 2.44%
Registros de alta geração eólica: 245
Percentual dentro da geração eólica: 2.87%


### 8. Comparação entre as fontes

In [ ]:
if pct_solar_alta5 > pct_eolica_alta5:
    print('A geração solar aparece com maior frequência acima de 70% do seu próprio máximo.')
elif pct_eolica_alta5 > pct_solar_alta5:
    print('A geração eólica aparece com maior frequência acima de 70% do seu próprio máximo.')
else:
    print('As duas fontes aparecem com a mesma frequência acima de 70% do máximo.')

A geração eólica aparece com maior frequência acima de 70% do seu próprio máximo.


### 9. Interpretação

Na amostra analisada, a geração solar apresentou valor máximo de 16.316,
enquanto a geração eólica apresentou máximo de 22.929. Por possuírem escalas
diferentes, foram calculados limiares próprios de 70%: 11.421,20 para Solar e
16.050,30 para Wind.

Foram encontrados 45 registros de alta geração solar, equivalentes a 2,44%
dos registros solares. Para a geração eólica, foram encontrados 245 registros
de alta geração, equivalentes a 2,87% dos registros eólicos.

Portanto, a geração eólica apareceu com maior frequência acima de 70% do seu
próprio máximo. Não seria adequado aplicar o mesmo limite numérico às duas
fontes, pois a escala máxima da geração eólica é maior que a solar. Um único
limite poderia tornar a comparação injusta, selecionando uma quantidade
desproporcional de registros de uma das fontes.

# **Dataset 6 — Individual Household Electric Power Consumption**

**Situação:** identificar episódios de demanda elevada que também apresentam
corrente elétrica acima do comportamento médio da residência.

### 1. Carregar a amostra e inspecionar

In [ ]:
url6 = '/content/Amostra_base_Individual_Household_Electric_Power_Consumption.csv'

df6 = pd.read_csv(url6)

df6.head()

,Date,Time,Global_active_power,Global_reactive_power,Voltage,Global_intensity,Sub_metering_1,Sub_metering_2,Sub_metering_3
0,7/7/2010,18:10:00,0.256,0.106,242.00,1.2,0.0,0.0,1.0
1,14/5/2007,06:50:00,0.466,0.352,237.22,2.4,0.0,2.0,0.0
2,26/9/2007,18:10:00,0.758,0.194,238.66,3.2,0.0,1.0,0.0
3,19/6/2007,07:30:00,1.290,0.046,240.64,5.4,1.0,0.0,18.0
4,10/5/2010,04:43:00,0.428,0.202,242.23,1.8,0.0,2.0,1.0


In [ ]:
print(f'Total de registros: {df6.shape[0]}')
print(f'Total de colunas: {df6.shape[1]}')

Total de registros: 207526
Total de colunas: 9


In [ ]:
df6.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 207526 entries, 0 to 207525
Data columns (total 9 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   Date                   207526 non-null  object 
 1   Time                   207526 non-null  object 
 2   Global_active_power    207526 non-null  float64
 3   Global_reactive_power  207526 non-null  float64
 4   Voltage                207526 non-null  float64
 5   Global_intensity       207526 non-null  float64
 6   Sub_metering_1         207526 non-null  float64
 7   Sub_metering_2         207526 non-null  float64
 8   Sub_metering_3         207526 non-null  float64
dtypes: float64(7), object(2)
memory usage: 14.2+ MB


In [ ]:
df6.describe()

,Global_active_power,Global_reactive_power,Voltage,Global_intensity,Sub_metering_1,Sub_metering_2,Sub_metering_3
count,207526.000000,207526.000000,207526.000000,207526.000000,207526.000000,207526.000000,207526.000000
mean,1.093805,0.124063,240.843654,4.637489,1.134004,1.293732,6.486336
std,1.052250,0.112334,3.216827,4.423019,6.158917,5.793760,8.386876
min,0.078000,0.000000,224.670000,0.200000,0.000000,0.000000,0.000000
25%,0.310000,0.050000,239.040000,1.400000,0.000000,0.000000,0.000000
50%,0.632000,0.102000,240.970000,2.800000,0.000000,0.000000,1.000000
75%,1.522000,0.194000,242.860000,6.400000,0.000000,1.000000,17.000000
max,10.650000,1.390000,253.750000,46.400000,88.000000,77.000000,31.000000


### 2. Organizar e simplificar os nomes das colunas

In [ ]:
df6 = df6.rename(columns={
    'Date': 'Data',
    'Time': 'Hora',
    'Global_active_power': 'Potencia_Ativa',
    'Global_reactive_power': 'Potencia_Reativa',
    'Voltage': 'Tensao',
    'Global_intensity': 'Corrente',
    'Sub_metering_1': 'Submedicao_1',
    'Sub_metering_2': 'Submedicao_2',
    'Sub_metering_3': 'Submedicao_3'
})

df6.columns

Index(['Data', 'Hora', 'Potencia_Ativa', 'Potencia_Reativa', 'Tensao',
       'Corrente', 'Submedicao_1', 'Submedicao_2', 'Submedicao_3'],
      dtype='object')

### 3. Maior potência ativa e limiar de 75%

In [ ]:
potencia_max6 = df6['Potencia_Ativa'].max()
limiar6 = 0.75 * potencia_max6

print(f'Maior potência ativa: {potencia_max6:.2f}')
print(f'Limiar de 75%: {limiar6:.2f}')

Maior potência ativa: 10.65
Limiar de 75%: 7.99


### 4. DataFrame de demanda elevada, quantidade e percentual

In [ ]:
df6_potencia_alta = df6[
    df6['Potencia_Ativa'] > limiar6
]

qtd_potencia_alta6 = df6_potencia_alta.shape[0]

pct_potencia_alta6 = (
    qtd_potencia_alta6 / df6.shape[0] * 100
)

print(f'Registros acima do limiar: {qtd_potencia_alta6}')
print(f'Percentual da amostra: {pct_potencia_alta6:.2f}%')

df6_potencia_alta.head()

Registros acima do limiar: 33
Percentual da amostra: 0.02%


,Data,Hora,Potencia_Ativa,Potencia_Reativa,Tensao,Corrente,Submedicao_1,Submedicao_2,Submedicao_3
24635,21/3/2008,15:52:00,8.110,0.196,236.87,34.2,37.0,74.0,17.0
42771,15/1/2008,12:26:00,8.026,0.058,234.91,34.0,37.0,73.0,17.0
45962,25/1/2009,13:27:00,8.218,0.336,236.05,34.8,37.0,73.0,17.0
46828,24/11/2009,19:39:00,9.708,0.372,231.37,42.0,37.0,69.0,17.0
55667,2/2/2008,19:46:00,8.168,0.540,228.91,35.6,35.0,35.0,17.0


### 5. Corrente média da amostra

In [ ]:
corrente_media6 = df6['Corrente'].mean()

print(f'Corrente média: {corrente_media6:.2f}')

Corrente média: 4.64


### 6. Potência ativa elevada e corrente acima da média

In [ ]:
df6_potencia_alta_corrente = df6[
    (df6['Potencia_Ativa'] > limiar6) &
    (df6['Corrente'] > corrente_media6)
]

print(
    'Registros com potência elevada e corrente acima da média:',
    df6_potencia_alta_corrente.shape[0]
)

Registros com potência elevada e corrente acima da média: 33


### 7. Comparação entre os dois DataFrames

In [ ]:
qtd_potencia_corrente6 = (
    df6_potencia_alta_corrente.shape[0]
)

reducao6 = qtd_potencia_alta6 - qtd_potencia_corrente6

print(f'Somente potência ativa elevada: {qtd_potencia_alta6}')
print(
    'Potência ativa elevada e corrente acima da média:',
    qtd_potencia_corrente6
)
print(f'Redução de registros: {reducao6}')

Somente potência ativa elevada: 33
Potência ativa elevada e corrente acima da média: 33
Redução de registros: 0


### 8. Interpretação

A maior potência ativa registrada na amostra foi de 10,65. O limiar de demanda
elevada, calculado como 75% desse máximo, foi de 7,99. Foram encontrados 33
registros acima desse limite, o que representa aproximadamente 0,02% da
amostra.

A corrente média da residência foi de 4,64. Ao adicionar a condição de corrente
acima da média, a quantidade de registros permaneceu em 33. Portanto, não houve
redução no segundo DataFrame.

Isso indica que todos os episódios de potência ativa muito elevada da amostra
também ocorreram com corrente acima do comportamento médio. Nesse caso, a
segunda condição confirma a associação entre alta demanda e corrente elevada,
em vez de excluir parte dos registros.